File: 20-actors/matsurigoto/methods/modules/test_corp_registry.py
#!/usr/bin/env python3
"""Conformance tests for the corp-registry module (matsurigoto 政, ADR-2606052300).

The LEI tests exercise the real ISO 7064 MOD 97-10 checksum: an assembled LEI validates (mod
97 == 1, proven algebraically), and flipping any character breaks it. Standalone + pytest.
"""
from __future__ import annotations

import pathlib
import sys

HERE = pathlib.Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))

import corp_registry as R  # noqa: E402


def test_no_server_authority_certificate_unsigned():
    """G1 — certificate unsigned, module signs nothing."""
    assert R.SERVER_HELD_AUTHORITY is False
    r = R.register_incorporation("Co", ["o"], 0, "art", "addr", "JPN", 1)
    assert r["certificate"]["proof"] is None
    assert r["certificate"]["server_held_authority"] is False


def test_to_digits_iso7064_mapping():
    """A=10 … Z=35; digits pass through."""
    assert R._to_digits("0A9Z") == "0" + "10" + "9" + "35"


def test_lei_roundtrip_validates():
    """An assembled LEI (check digits computed) must satisfy mod 97 == 1."""
    lei = R.assign_lei("EZHY", "000000000001")
    assert len(lei) == 20
    assert R.validate_lei(lei) is True


def test_lei_check_digits_make_mod97_one_for_many_entities():
    """Property holds for many distinct entity ids (exercises the MOD-97-10 arithmetic)."""
    for n in range(1, 50):
        lei = R.assign_lei("EZHY", f"{n:012d}")
        assert R.validate_lei(lei), lei


def test_lei_corruption_detected():
    """Flipping a character breaks the checksum (the test that proves digit-dependence)."""
    lei = R.assign_lei("EZHY", "000000000042")
    assert R.validate_lei(lei)
    # mutate one char in the entity-id region
    bad = list(lei)
    bad[8] = "Z" if bad[8] != "Z" else "Y"
    assert R.validate_lei("".join(bad)) is False


def test_lei_rejects_bad_length_and_chars():
    assert R.validate_lei("TOOSHORT") is False
    assert R.validate_lei("EZHY00000000000001*9") is False  # '*' illegal
    assert R.validate_lei(12345) is False


def test_check_digits_two_chars_zero_padded():
    cd = R.compute_lei_check_digits("EZHY00" + "000000000007")
    assert len(cd) == 2 and cd.isdigit()


def test_incorporation_assigns_registry_number_and_lei():
    r = R.register_incorporation("Tree of Life K.K.", ["officer:rin"], 10_000_000,
                                 "articles", "東京都", "JPN", 7)
    assert r["registry_number"] == "JPN-00000007"
    assert R.validate_lei(r["lei"])
    assert r["record"]["immutable"] is True


def test_incorporation_validation_rules():
    bad_args = [
        ("", ["o"], 0, "a", "ad"),       # no name
        ("Co", [], 0, "a", "ad"),        # no officer
        ("Co", ["o"], -1, "a", "ad"),    # negative capital
        ("Co", ["o"], 0, "", "ad"),      # no articles
        ("Co", ["o"], 0, "a", ""),       # no address
    ]
    for args in bad_args:
        try:
            R.register_incorporation(*args, "JPN", 1)
        except ValueError:
            continue
        raise AssertionError(f"should reject {args}")


def test_change_is_append_only_g5():
    hist = []
    inc = R.register_incorporation("Co", ["o"], 0, "a", "ad", "JPN", 1)
    hist = R.append(hist, inc)
    chg = R.register_change(inc["registry_number"], {"address": "new"}, "2026-06-05")
    hist2 = R.append(hist, chg)
    assert len(hist) == 1 and len(hist2) == 2          # original untouched, new list
    assert hist2[0]["kind"] == "incorporation"          # incorporation record preserved
    assert hist2[1]["kind"] == "change"                 # amendment appended


def test_solve_is_gated_at_r0():
    try:
        R.solve()
    except RuntimeError:
        return
    raise AssertionError("solve() must raise at R0")


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
        print(f"  ok  {fn.__name__}")
    print(f"\n{len(fns)}/{len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/matsurigoto/methods/modules/test_corp_registry.clj
;; test_corp_registry.clj — matsurigoto corp-registry: ISO 7064 MOD 97-10 LEI parity with
;; corp_registry.py + append-only registry discipline. Run via `bb test:matsurigoto`. ADR-2606142300.
(ns matsurigoto.methods.modules.test-corp-registry
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [matsurigoto.methods.modules.corp-registry :as cr]))

(deftest iso7064-lei-parity
  (testing "ISO 17442 LEI + ISO 7064 MOD 97-10 reproduced exactly (golden from corp_registry.py)"
    (is (= "68" (cr/compute-lei-check-digits "EZHY00000000000001")))
    (is (= "EZHY0000000000000168" (cr/assign-lei "EZHY" "000000000001")))
    (is (= "549300ACME0000000169" (cr/assign-lei "5493" "ACME00000001")))
    (is (true? (cr/validate-lei "EZHY0000000000000168")))
    (is (true? (cr/validate-lei "549300ACME0000000169")))))

(deftest lei-validation-rejects-corruption
  (testing "flipping any character breaks the MOD 97-10 check (prob 96/97)"
    (is (false? (cr/validate-lei "EZHY0000000000000169")))   ; last digit flipped
    (is (false? (cr/validate-lei "EZHY000000000000016")))    ; wrong length (19)
    (is (false? (cr/validate-lei "EZHY00000000000001688")))  ; wrong length (21)
    (is (false? (cr/validate-lei nil)))
    (is (false? (cr/validate-lei "EZHY000000000000016!")))))  ; non [0-9A-Z]

(deftest assign-lei-guards
  (testing "LEI assembly input guards"
    (is (thrown? Exception (cr/assign-lei "EZH" "000000000001")))    ; LOU not 4
    (is (thrown? Exception (cr/assign-lei "EZHY" "0001")))           ; entity not 12
    (is (thrown? Exception (cr/compute-lei-check-digits "TOOSHORT")))))

(deftest incorporation-record
  (testing "incorporation assigns registry number + valid LEI (golden)"
    (let [r (cr/register-incorporation {:entity-name "Tree of Life K.K." :officers ["officer:rin"]
                                        :capital 10000000 :articles "articles-hash"
                                        :address "東京都新宿区" :jurisdiction "JPN" :sequence 1})]
      (is (= "JPN-00000001" (:registry-number r)))
      (is (= "EZHY0000000000000168" (:lei r)))
      (is (true? (cr/validate-lei (:lei r))))
      (is (true? (:immutable (:record r))))                  ; G5
      (let [c (:certificate r)]
        (is (nil? (:proof c)))                               ; G1
        (is (false? (:server-held-authority c)))             ; G1
        (is (= "issued-unsigned" (:status c)))
        (is (= ["VerifiableCredential" "IncorporationCertificate"] (:type c)))))))

(deftest incorporation-guards
  (testing "incorporation validation guards"
    (let [base {:entity-name "X" :officers ["o"] :capital 0 :articles "a" :address "addr"
                :jurisdiction "JPN" :sequence 0}]
      (is (map? (cr/register-incorporation base)))            ; valid baseline
      (is (thrown? Exception (cr/register-incorporation (assoc base :entity-name ""))))
      (is (thrown? Exception (cr/register-incorporation (assoc base :officers []))))
      (is (thrown? Exception (cr/register-incorporation (assoc base :capital -1))))
      (is (thrown? Exception (cr/register-incorporation (assoc base :sequence -1)))))))

(deftest append-only-change
  (testing "G5 — a change is an appended amendment, never an overwrite"
    (let [inc (cr/register-incorporation {:entity-name "X" :officers ["o"] :capital 0 :articles "a"
                                          :address "addr" :jurisdiction "JPN" :sequence 1})
          chg (cr/register-change "JPN-00000001" {:address "大阪市"} "2026-06-15")
          hist (-> [] (cr/append inc) (cr/append chg))]
      (is (= "JPN-00000001#chg@2026-06-15" (:record-id (:record chg))))
      (is (= "change" (:kind (:record chg))))
      (is (= 2 (count hist)))                                 ; append-only: both records retained
      (is (= "incorporation" (:kind (first hist))))
      (is (true? (:immutable (:record chg))))
      (is (thrown? Exception (cr/register-change "" {:a 1} "2026-06-15")))
      (is (thrown? Exception (cr/register-change "JPN-1" {} "2026-06-15"))))))

(deftest live-registration-gated
  (testing "G1 — live registration against a real register is Council+operator gated"
    (is (thrown? Exception (cr/solve)))
    (is (false? cr/server-held-authority))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'matsurigoto.methods.modules.test-corp-registry)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
